File: internal/tools/grep.go
package tools

import (
	"bufio"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type GrepTool struct{}

func (GrepTool) Name() string {
	return "grep"
}

func (GrepTool) Description() string {
	return "Search file contents using regex patterns. Returns matching lines with file paths and line numbers."
}

func (GrepTool) Parameters() map[string]any {
	return objectSchema([]string{"pattern"}, map[string]any{
		"pattern": stringSchema("The regex pattern to search for"),
		"path":    stringSchema("File or directory to search in (defaults to cwd)"),
		"include": stringSchema("Glob pattern to filter files (e.g. '*.ts')"),
	})
}

func (GrepTool) Call(input map[string]any) Result {
	pattern, err := stringArg(input, "pattern")
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}
	searchPath, ok, err := optionalStringArg(input, "path")
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}
	if !ok || searchPath == "" {
		searchPath = "."
	}
	include, includeOK, err := optionalStringArg(input, "include")
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}

	expression, err := regexp.Compile(pattern)
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}

	matches, err := grepFiles(expression, searchPath, include, includeOK)
	if err != nil {
		return Result{Output: "Error: " + err.Error(), IsError: true}
	}
	if len(matches) == 0 {
		return Result{Output: "No matches found."}
	}
	return Result{Output: strings.Join(matches, "\n")}
}

func (GrepTool) RenderToolCall(input map[string]any) string {
	pattern, _ := stringArg(input, "pattern")
	searchPath, pathOK, _ := optionalStringArg(input, "path")
	include, includeOK, _ := optionalStringArg(input, "include")

	description := toolName("grep") + " " + muted("/") + pattern + muted("/")
	if pathOK && searchPath != "" {
		description += " " + muted("in") + " " + fileName(searchPath)
	}
	if includeOK && include != "" {
		description += " " + muted("("+include+")")
	}
	return description
}

func (GrepTool) RenderResult(result Result) string {
	if result.IsError {
		return failure(result.Output)
	}
	return truncateLines(result.Output, 20, "more lines")
}

func grepFiles(expression *regexp.Regexp, searchPath, include string, includeOK bool) ([]string, error) {
	info, err := os.Stat(searchPath)
	if err != nil {
		return nil, err
	}

	var files []string
	if !info.IsDir() {
		files = append(files, searchPath)
	} else {
		err := filepath.WalkDir(searchPath, func(current string, entry fs.DirEntry, walkErr error) error {
			if walkErr != nil {
				return nil
			}
			if entry.IsDir() {
				return nil
			}
			files = append(files, current)
			return nil
		})
		if err != nil {
			return nil, err
		}
	}

	var matches []string
	for _, filePath := range files {
		if includeOK && include != "" {
			ok, err := filepath.Match(include, filepath.Base(filePath))
			if err != nil {
				return nil, err
			}
			if !ok {
				continue
			}
		}
		fileMatches := grepFile(expression, filePath, 100-len(matches))
		matches = append(matches, fileMatches...)
		if len(matches) >= 100 {
			break
		}
	}
	return matches, nil
}

func grepFile(expression *regexp.Regexp, filePath string, remaining int) []string {
	if remaining <= 0 {
		return nil
	}
	file, err := os.Open(filePath)
	if err != nil {
		return nil
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 1024), 1024*1024)

	var matches []string
	lineNumber := 0
	for scanner.Scan() {
		lineNumber++
		line := scanner.Text()
		if !expression.MatchString(line) {
			continue
		}
		if len(line) > 200 {
			line = line[:200]
		}
		matches = append(matches, fmt.Sprintf("%s:%d:%s", filePath, lineNumber, line))
		if len(matches) >= remaining {
			break
		}
	}
	return matches
}

File: kai/tools/grep_.py
from __future__ import annotations

import fnmatch
import os
import re
from typing import Any

from kai.tools.base import (
    Result,
    Tool,
    object_schema,
    optional_string_arg,
    string_schema,
)
from kai.tools.colors import failure, file_name, muted, tool_name
from kai.tools.format import truncate_lines


class GrepTool(Tool):
    def name(self) -> str:
        return "grep"

    def description(self) -> str:
        return "Search file contents using regex patterns. Returns matching lines with file paths and line numbers."

    def parameters(self) -> dict[str, Any]:
        return object_schema(
            ["pattern"],
            {
                "pattern": string_schema("The regex pattern to search for"),
                "path": string_schema(
                    "File or directory to search in (defaults to cwd)"
                ),
                "include": string_schema("Glob pattern to filter files (e.g. '*.ts')"),
            },
        )

    def call(self, input: dict[str, Any]) -> Result:
        pattern = input.get("pattern")
        if not isinstance(pattern, str) or not pattern:
            return Result(
                output="Error: missing required argument 'pattern'", is_error=True
            )

        search_path, _ = optional_string_arg(input, "path")
        if not search_path:
            search_path = "."

        include, include_specified = optional_string_arg(input, "include")

        try:
            expression = re.compile(pattern)
        except re.error as e:
            return Result(output=f"Error: {e}", is_error=True)

        try:
            matches = _grep_files(
                expression, search_path, include if include_specified else None
            )
        except Exception as e:
            return Result(output=f"Error: {e}", is_error=True)

        if not matches:
            return Result(output="No matches found.")
        return Result(output="\n".join(matches))

    def render_tool_call(self, input: dict[str, Any]) -> str:
        pattern = input.get("pattern", "")
        search_path, path_specified = optional_string_arg(input, "path")
        include, include_specified = optional_string_arg(input, "include")

        description = f"{tool_name('grep')} {muted('/')}{pattern}{muted('/')}"
        if path_specified and search_path:
            description += f" {muted('in')} {file_name(search_path)}"
        if include_specified and include:
            description += f" {muted(f'({include})')}"
        return description

    def render_result(self, result: Result) -> str:
        if result.is_error:
            return failure(result.output)
        return truncate_lines(result.output, 20, "more lines")


def _grep_files(
    expression: re.Pattern[str],
    search_path: str,
    include: str | None,
) -> list[str]:
    if os.path.isfile(search_path):
        files = [search_path]
    else:
        files = []
        for dirpath, _dirnames, filenames in os.walk(search_path):
            for filename in filenames:
                files.append(os.path.join(dirpath, filename))

    matches: list[str] = []
    for file_path in files:
        if include is not None and not fnmatch.fnmatch(
            os.path.basename(file_path), include
        ):
            continue
        file_matches = _grep_file(expression, file_path, 100 - len(matches))
        matches.extend(file_matches)
        if len(matches) >= 100:
            break

    return matches


def _grep_file(
    expression: re.Pattern[str],
    file_path: str,
    remaining: int,
) -> list[str]:
    if remaining <= 0:
        return []
    try:
        with open(file_path, errors="replace") as f:
            lines = f.readlines()
    except OSError:
        return []

    matches: list[str] = []
    for line_number, line in enumerate(lines, 1):
        if not expression.search(line):
            continue
        display_line = line.rstrip("\n\r")
        if len(display_line) > 200:
            display_line = display_line[:200]
        matches.append(f"{file_path}:{line_number}:{display_line}")
        if len(matches) >= remaining:
            break
    return matches
